File: frontend/apps/ios/Pantopus/Core/Auth/DPoPProofBuilder.swift
//
//  DPoPProofBuilder.swift
//  Pantopus
//
//  Builds the `DPoP: <jwt>` proof (RFC 9449, ES256) that binds a request to
//  the device key. Wire shape pinned by docs/persistent-login/CONTRACT.md:
//
//    header : {"typ":"dpop+jwt","alg":"ES256","jwk":{kty,crv,x,y}}
//    payload: {"jti":uuid,"htm":"POST","htu":"<scheme>://<host>[:port]<path>",
//              "iat":<unix s>,"rth"?:base64url(sha256(refreshToken))}
//    sig    : raw r||s (64 bytes), base64url
//
//  `rth` is REQUIRED on `/api/users/refresh` and `/api/users/logout`
//  whenever a refresh token travels in the body — the proof is then
//  useless without that exact token and vice versa.
//

import CryptoKit
import Foundation

enum DPoPProofBuilder {
    enum BuildError: Error, Equatable {
        case invalidURL
    }

    /// Canonical `htu`: scheme + host (+ explicit port) + path, no query, no
    /// fragment. The server compares against `PUBLIC_API_BASE_URL + path`
    /// (or `<proto>://<host>` from the request when unset), so the client's
    /// own base URL is the right thing to hash — proxies do not matter.
    static func htu(for url: URL) -> String? {
        guard let scheme = url.scheme?.lowercased(), let host = url.host?.lowercased() else {
            return nil
        }
        let port = url.port.map { ":\($0)" } ?? ""
        let path = url.path.isEmpty ? "/" : url.path
        return "\(scheme)://\(host)\(port)\(path)"
    }

    /// `base64url(sha256(refreshToken))` — the `rth` claim.
    static func refreshTokenHash(_ refreshToken: String) -> String {
        Base64URL.encode(Data(SHA256.hash(data: Data(refreshToken.utf8))))
    }

    /// Assemble and sign a proof.
    ///
    /// - Parameters:
    ///   - signer: the device key (or a software key in tests).
    ///   - method: HTTP method, upper-cased on the wire.
    ///   - url: the full request URL; only scheme/host/port/path are used.
    ///   - refreshToken: when present, adds `rth`.
    ///   - jti / now: injectable for deterministic tests.
    static func build(
        signer: some DPoPSigner,
        method: String,
        url: URL,
        refreshToken: String? = nil,
        jti: UUID = UUID(),
        now: Date = Date()
    ) throws -> String {
        guard let htu = htu(for: url) else { throw BuildError.invalidURL }
        let header = Header(jwk: signer.jwk)
        let payload = Payload(
            jti: jti.uuidString.lowercased(),
            htm: method.uppercased(),
            htu: htu,
            iat: Int(now.timeIntervalSince1970),
            rth: refreshToken.map(refreshTokenHash)
        )
        let encoder = JSONEncoder()
        // Slashes must not be escaped (`\/`) — the server tolerates it, but
        // a stable byte layout keeps test vectors readable.
        encoder.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
        let headerPart = try Base64URL.encode(encoder.encode(header))
        let payloadPart = try Base64URL.encode(encoder.encode(payload))
        let signingInput = "\(headerPart).\(payloadPart)"
        let signature = try signer.sign(Data(signingInput.utf8))
        return "\(signingInput).\(Base64URL.encode(signature))"
    }

    // MARK: - JOSE structures

    struct Header: Encodable {
        let typ = "dpop+jwt"
        let alg = "ES256"
        let jwk: JWK
    }

    struct Payload: Codable, Equatable {
        let jti: String
        let htm: String
        let htu: String
        let iat: Int
        let rth: String?
    }

    /// The three decoded segments of a compact JWT.
    struct Parts: Equatable {
        let header: Data
        let payload: Data
        let signature: Data
    }

    /// Split a compact JWT into its decoded parts — for tests and
    /// diagnostics only (the client never verifies its own proofs).
    static func decodeParts(_ jwt: String) -> Parts? {
        let parts = jwt.split(separator: ".", omittingEmptySubsequences: false)
        guard parts.count == 3,
              let header = Base64URL.decode(String(parts[0])),
              let payload = Base64URL.decode(String(parts[1])),
              let signature = Base64URL.decode(String(parts[2])) else {
            return nil
        }
        return Parts(header: header, payload: payload, signature: signature)
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/auth/DPoPProofBuilder.kt
package app.pantopus.android.data.auth

import com.squareup.moshi.JsonAdapter
import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Builds the `DPoP: <jwt>` proof (RFC 9449 profile pinned by
 * `docs/persistent-login/CONTRACT.md` §Headers):
 *
 * ```
 * header : {"typ":"dpop+jwt","alg":"ES256","jwk":{"kty":"EC","crv":"P-256","x":…,"y":…}}
 * payload: {"jti":"<uuid>","htm":"POST","htu":"<scheme>://<host>[:port]<path>","iat":<unix s>,
 *           "rth":"<b64url sha256(refreshToken)>"}          // only when a refresh token is sent
 * signature: raw r||s (64 bytes) base64url
 * ```
 *
 * `htu` never carries a query string. The server compares it against
 * `PUBLIC_API_BASE_URL + path` (or `<proto>://<host>` derived from the
 * request when that env is unset), so we build it from the same base URL
 * Retrofit uses ([htu]). `iat` is accepted within +/-300 s; `jti` is
 * single-use for 10 min — every proof is minted fresh, never reused across
 * retries (see `TokenAuthenticator`, which re-mints on replay).
 *
 * Pure JVM (no `android.util.Base64`, no `org.json`) so it is unit-testable
 * with a software EC key.
 */
@Singleton
class DPoPProofBuilder
    @Inject
    constructor() {
        private val jsonAdapter: JsonAdapter<Map<String, Any>> =
            Moshi
                .Builder()
                .build()
                .adapter(Types.newParameterizedType(Map::class.java, String::class.java, Any::class.java))

        /**
         * Mint a proof for `htm htu` signed by [key].
         *
         * @param refreshToken when non-null, `rth = base64url(sha256(refreshToken))`
         *   is added — REQUIRED on `/api/users/refresh` and `/api/users/logout`
         *   whenever a refresh token travels in the body.
         * @param nowSeconds / [jti] injectable for deterministic tests.
         */
        fun build(
            key: DeviceSigningKey,
            htm: String,
            htu: String,
            refreshToken: String? = null,
            nowSeconds: Long = System.currentTimeMillis() / MILLIS_PER_SECOND,
            jti: String = UUID.randomUUID().toString(),
        ): String {
            val header =
                linkedMapOf<String, Any>(
                    "typ" to TYP,
                    "alg" to ALG,
                    "jwk" to key.jwk,
                )
            val payload =
                linkedMapOf<String, Any>(
                    "jti" to jti,
                    "htm" to htm.uppercase(),
                    "htu" to htu,
                    "iat" to nowSeconds,
                )
            if (refreshToken != null) payload["rth"] = refreshTokenHash(refreshToken)
            val signingInput =
                EcKeyCodec.base64Url(jsonAdapter.toJson(header).toByteArray(Charsets.UTF_8)) +
                    "." +
                    EcKeyCodec.base64Url(jsonAdapter.toJson(payload).toByteArray(Charsets.UTF_8))
            val signature = key.sign(signingInput.toByteArray(Charsets.US_ASCII))
            return signingInput + "." + EcKeyCodec.base64Url(signature)
        }

        companion object {
            const val TYP = "dpop+jwt"
            const val ALG = "ES256"
            private const val MILLIS_PER_SECOND = 1000L

            /** `base64url(sha256(refreshToken))` — the `rth` claim. */
            fun refreshTokenHash(refreshToken: String): String =
                EcKeyCodec.base64Url(EcKeyCodec.sha256(refreshToken.toByteArray(Charsets.UTF_8)))

            /**
             * `<scheme>://<host>[:port]<path>` for [path] under [baseUrl], with
             * any query / fragment dropped and the default port elided — the
             * exact form the backend reconstructs. [path] may be given with
             * or without a leading slash (Retrofit-style relative paths).
             */
            fun htu(
                baseUrl: String,
                path: String,
            ): String {
                val base: HttpUrl =
                    requireNotNull((baseUrl.trimEnd('/') + "/").toHttpUrlOrNull()) { "Invalid API base URL: $baseUrl" }
                val cleanPath = "/" + path.substringBefore('?').substringBefore('#').trimStart('/')
                return htu(base.newBuilder().encodedPath(cleanPath).query(null).fragment(null).build())
            }

            /** [htu] for an already-resolved request URL (query + fragment dropped). */
            fun htu(url: HttpUrl): String {
                val port = if (url.port == HttpUrl.defaultPort(url.scheme)) "" else ":${url.port}"
                return "${url.scheme}://${url.host}$port${url.encodedPath}"
            }
        }
    }
